firedrake/mg/interface.py

```diff
@@ -91,16 +112,37 @@ def inject(fine, coarse):
     hierarchy, lvl = utils.get_level(cfs)
     if hierarchy is None:
         raise RuntimeError("Coarse function not from hierarchy")
+    # for skipping hierarchies
+    if hasattr(hierarchy, '_full_hierarchy') == 1:
+        lvl = lvl * hierarchy._skip
+        hierarchy = hierarchy._full_hierarchy
     fhierarchy, flvl = utils.get_level(fine.function_space())
-    if flvl != lvl + 1:
-        raise ValueError("Can only inject from level %d to level %d, not %d" %
-                         (flvl, flvl - 1, lvl))
+    # for skipping hierarchies
+    if hasattr(fhierarchy, '_full_hierarchy') == 1:
+        flvl = flvl * fhierarchy._skip
+        fhierarchy = fhierarchy._full_hierarchy
+    if lvl > flvl:
+        raise ValueError("Cannot inject from level %d to level %d" %
+                         (flvl, lvl))
+    # if the same level, return current function
+    if lvl == flvl:
+        coarse.assign(fine)
+        return
+    # number of recursive injections
+    slvl = flvl - lvl
     if hierarchy is not fhierarchy:
         raise ValueError("Can't prolong between functions from different hierarchies")
     if isinstance(hierarchy, firedrake.MixedFunctionSpaceHierarchy):
         for f, c in zip(fine.split(), coarse.split()):
             inject(f, c)
         return
-    op2.par_loop(hierarchy._inject_kernel, hierarchy._cell_sets[lvl],
-                 coarse.dat(op2.WRITE, coarse.cell_node_map()[op2.i[0]]),
-                 fine.dat(op2.READ, hierarchy.cell_node_map(lvl)))
+    # carry out recursive injections
+    finer = Function(hierarchy[flvl]).assign(fine)
+    for j in range(slvl):
+        intermediate = Function(hierarchy[flvl - j - 1])
+        op2.par_loop(hierarchy._inject_kernel, hierarchy._cell_sets[flvl - j - 1],
+                     intermediate.dat(op2.WRITE, intermediate.cell_node_map()[op2.i[0]]),
+                     finer.dat(op2.READ, hierarchy.cell_node_map(flvl - j - 1)))
+        if j < slvl - 1:
+            finer = Function(hierarchy[flvl - j - 1]).assign(intermediate)
+    coarse.assign(intermediate)
```

Comment:
I have never used restriction before, and don't know how to use it correctly. Do you mind explaining then I will make the necessary changes.